futures-util/src/future/first_ok.rs

```diff
@@ -0,0 +1,220 @@
+use crate::future::{Fuse, FutureExt};
+use alloc::vec::Vec;
+use core::iter::FromIterator;
+use core::pin::Pin;
+use futures_core::future::{FusedFuture, Future, TryFuture};
+use futures_core::task::{Context, Poll};
+
+/// Future for the [`first_ok()`] function.
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[derive(Debug, Clone)]
+pub struct FirstOk<F> {
+    // Critical safety invariant: after FirstAll is created, this vector can
+    // never be reallocated, nor can its contents be moved, in order to ensure
+    // that Pin is upheld.
+    futures: Vec<F>,
+}
+
+// Safety: once created, the contents of the vector don't change, and they'll
+// remain in place permanently.
+impl<F> Unpin for FirstOk<F> {}
+
+impl<T, E, F> Future for FirstOk<F>
+where
+    F: Future<Output = Result<T, E>> + FusedFuture,
```

Comment:
I considered this, and if we remove FusedFuture entirely (like you suggested) then we can go this route, but I went with the trait here because many futures are "inherently" fused, so in the interest of zero-cost abstractions I thought it made more sense to just have the user call `.fuse()` only if they need to

Reply:
> so in the interest of zero-cost abstractions I thought it made more sense to just have the user call `.fuse()` only if they need to

Patterns like newtype are certainly not perfect zero-cost, but `Fuse` just adds a flag to indicate if the future is complete and I think the actual cost is very small.

Another reason I don't want to add new features that rely on `Fused*` traits is that they are very likely to be removed in the next major version. (see #2207)
Even if we add features that rely on `Fused*` traits, they will be changed to use `Fuse` or `MaybeDone`. And if users forget to remove `.fuse()` on update to new version, it will eventually be even more inefficient.